cuqi/distribution/_gaussian.py

```diff
@@ -163,37 +215,342 @@ def gradient(self, val, *args, **kwargs):
-        `s = mean + pseudoinverse(sqrtprec)*eps`,
-        where `eps` is a standard normal noise and `s`is the desired sample 
+        `s = mean + pseudoinverse(sqrtprec)*e`,
+        where `e` is a standard Gaussian random vector and `s` is the desired sample 
         """
-        #Sample N(0,I)
+        # Sample N(0,I)
         if rng is not None:
-            e = rng.randn(np.shape(self.sqrtprec)[0],N)
+            e = rng.randn(np.shape(self.sqrtprec)[0], N)
         else:
-            e = np.random.randn(np.shape(self.sqrtprec)[0],N)
-
-        #Compute permutation
-        if N==1: #Ensures we add (dim,1) with (dim,1) and not with (dim,)
-            permutation = splinalg.spsolve(self.sqrtprec,e)[:,None]
+            e = np.random.randn(np.shape(self.sqrtprec)[0], N)
+
+        # Compute perturbation
+        if spa.issparse(self.sqrtprec): # do sparse
+            # if np.allclose(self.sqrtprec, spa.tril(self.sqrtprec, format='csr')): # matrix is triangular
+            #     perturbation = spa.linalg.spsolve_triangular(self.sqrtprec, e)
+            if (N == 1):
+                perturbation = spa.linalg.spsolve(self.sqrtprec, e)[:, None]
+            else:
+                perturbation = spa.linalg.spsolve(self.sqrtprec, e)
         else:
-            permutation = splinalg.spsolve(self.sqrtprec,e)
-            
-        # Add to mean
-        s = self.mean[:,None] + permutation
+            if np.allclose(self.sqrtprec, np.tril(self.sqrtprec)): # matrix is triangular
+                perturbation = splinalg.solve_triangular(self.sqrtprec, e)
+            else:
+                perturbation = splinalg.solve(self.sqrtprec, e)
+
+        # Add mean
+        s = self.mean[:, None] + perturbation
         return s
 
+#===================================================
+#===================================================
+#===================================================
+def get_sqrtprec_from_cov(dim, cov, dimflag):
+    # cov is scalar, corrmat is identity or 1D
+    if (cov.shape[0] == 1): 
+        var = cov.ravel()[0]
+        logdet = dim*np.log(var)
+        rank = dim
+        if dimflag:
+            prec = (1/var)*spa.identity(dim, format="csr")
+            sqrtprec = np.sqrt(1/var)*spa.identity(dim, format="csr")
+        else:
+            prec = (1/var)*np.identity(dim)
+            sqrtprec = np.sqrt(1/var)*np.identity(dim)
+
+    # cov is vector
+    elif not spa.issparse(cov) and cov.shape[0] == np.size(cov): 
+        logdet = np.sum(np.log(cov))
+        rank = dim
+        if dimflag:
+            prec = spa.diags(1/cov, format="csr")
+            sqrtprec = spa.diags(np.sqrt(1/cov), format="csr")
+        else:
+            prec = np.diag(1/cov)
+            sqrtprec = np.diag(np.sqrt(1/cov))
+
+    # cov diagonal
+    elif (spa.issparse(cov) and cov.format == 'dia') or (not spa.issparse(cov) and np.count_nonzero(cov-np.diag(np.diagonal(cov))) == 0): 
```

Comment:
Somehow the "or" statement does not make sense to me. Since if cov is not sparse, we cannot use the .diagonal() method of sparse matrix anyway. Do we need the last check or can it be deleted?

Reply:
The first part checks for sparse diagonal matrices, and the second checks for dense diagonal ones. I will give it a little test